File: examples/governance/operations_domain/lib/operations_domain/deployment/commands/plan_deployment.rb
module OperationsDomain
  class Deployment
    module Commands
      class PlanDeployment
        include Hecks::Command
        emits "PlannedDeployment"

        attr_reader :model_id
        attr_reader :environment
        attr_reader :endpoint
        attr_reader :purpose
        attr_reader :audience

        def initialize(
          model_id: nil,
          environment: nil,
          endpoint: nil,
          purpose: nil,
          audience: nil
        )
          @model_id = model_id
          @environment = environment
          @endpoint = endpoint
          @purpose = purpose
          @audience = audience
        end

        def call
          Deployment.new(
            model_id: model_id,
            environment: environment,
            endpoint: endpoint,
            purpose: purpose,
            audience: audience,
            status: "planned"
          )
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/plan_deployment.go
package domain

import (
	"time"
)

type PlanDeployment struct {
	ModelId string `json:"model_id"`
	Environment string `json:"environment"`
	Endpoint string `json:"endpoint"`
	Purpose string `json:"purpose"`
	Audience string `json:"audience"`
}

func (c PlanDeployment) CommandName() string { return "PlanDeployment" }

func (c PlanDeployment) Execute(repo DeploymentRepository) (*Deployment, *PlannedDeployment, error) {
	agg := NewDeployment(c.ModelId, c.Environment, c.Endpoint, c.Purpose, c.Audience, time.Time{}, time.Time{}, "")
	agg.Status = "planned"
	if err := agg.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(agg); err != nil {
		return nil, nil, err
	}
	event := PlannedDeployment{
		AggregateID: agg.ID,
		ModelId: c.ModelId,
		Environment: c.Environment,
		Endpoint: c.Endpoint,
		Purpose: c.Purpose,
		Audience: c.Audience,
		OccurredAt: time.Now(),
	}
	return agg, &event, nil
}
